Add FMix Preprocessing Layer

quantumalaviya opened the pull request:
closes #254 

Adding support for FMix.

Updated image:
![fmix](https://user-images.githubusercontent.com/12191303/163289029-c2a2f934-61c4-48d9-b5e4-0e76500cc9fb.png)

quantumalaviya commented:
These changes are inspired by the MixUp layer and the official implementation. The functions are ported to Tensorflow for support. 
Also, currently, a batch will use very similar masks. This was particularly important to ensure the same intermediate tensor shapes and types (for vectorization)

quantumalaviya commented:
@LukeWood This took a bit longer than expected. Please review this when you get a chance!

LukeWood commented:
Wow!  This was faster than I expected.  Will review @quantumalaviya Thanks!

LukeWood commented:
A high level question... I wonder if the name `FourierCutMix` is more indicative of what is going on here.

LukeWood commented:
Or `FourierMix`

innat commented:
@LukeWood regarding the name, I think it's better to stick with the official name. We can think surely set a more indicative name but that might show disrespect to the official one. Also, the name `FMix` is known to many but `FourierMix`, people may get confused.

innat commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,207 @@
+        return sample_alpha / (sample_alpha + sample_beta)
+
+    def fftfreq(self, n, d=1):
+        # An implementation of numpy.fft.fftfreq using Tensorflow.
```
I think you should remove this line or give a reference. Officially they use numpy.

quantumalaviya replied:
Actually, I changed it to what the function does. This is more descriptive as well!

quantumalaviya commented:
I did consider naming it something more descriptive like FourierMix but decided to stick to `FMix` because there were no mentions of this in the paper. 
I do agree with @innat here but I can change it if that's important.

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,207 @@
+            distribution.  This controls the shape of the distribution from which the
+            smoothing values are sampled.  Defaults 0.5, which is a recommended value
+            in the paper.
+        decay_power: A float value representing the decay power for frequency decay prop
```
Lets remove abbreviations (unless prop isn't an abbreviation!)

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,207 @@
+            in the paper.
+        decay_power: A float value representing the decay power for frequency decay prop
+            1/f**d.  Defaults to 3, as recommended in the paper.
+        max_soft: Float value representing softening value between 0 and 0.5 which
```
could we rename this to `softening` or `edge_softening`

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,207 @@
+        sample_beta = tf.random.gamma(shape, 1.0, beta=beta)
+        return sample_alpha / (sample_alpha + sample_beta)
+
+    def fftfreq(self, n, d=1):
```
This can be static: @staticmethod

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,207 @@
+    def fftfreq(self, n, d=1):
+        # This function returns the sample frequencies of a discrete fourier transform.
+
+        N = tf.convert_to_tensor((n - 1) / 2 + 1)
```
Rename `N` to something else to avoid upper casing.  Also will make code less confusing.  Basically it looks like `N` is sampling the number of frequencies required to capture all information in an FFT (because the result of the FFT is mirrored).

Do we need the convert to tensor call?

quantumalaviya replied:
Added this as part of the range itself.

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,207 @@
+
+        N = tf.convert_to_tensor((n - 1) / 2 + 1)
+        results = tf.concat(
+            [tf.range(N, dtype=tf.int32), tf.range(-(n // 2), 0, dtype=tf.int32)], 0
```
I am a little confused what exactly is going on in this part.  If you don't know and this is taken from the original implementation, that is OK but if you DO know what is going on please leave some comments.

Why don't you add a docstring for this that follows what is in numpy's?

quantumalaviya replied:
I added some comments. Please have a look and lmk if that is descriptive enough.

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,207 @@
+        return results / (n * d)
+
+    def apply_fftfreq(self, h, w):
+        # Applying the fourier transform across 3 dimensions.
```
What's the third dimension?  Channels?

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,207 @@
+        )
+
+        param_size = tf.concat(
+            [tf.constant([ch]), tf.shape(freqs), tf.constant([2])], 0
```
lets expand some of the abbreviations where possible.  `channel` is easier to read than `ch`

LukeWood commented:
Wow!  Didn't realize how complex the signal processing portion of this code was!  Very cool, I reviewed as much as I can - there could be some value in adding extra comments to the signal processing functions as signal processing knowledge in the deep learning domain is pretty rare.  Thanks!

quantumalaviya commented:
I have added some comments and changed a few more variable names. Each function already has a small line describing what it does. I assumed that's enough context for anyone looking? A detailed description is available in the paper, which I added as part of the reference!

innat commented:
cc. @MattPainter01
Mentioning matt, who is one of the authors of FMix and officially implemented the `tf` version of the `f-mix`. It would be great to have some feedback from him.

bhack commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,221 @@
+        inputs["images"] = images
+        return inputs
+
+    def _augment(self, inputs):
```
Is this the only layer in KLPs where we have this behaviour?

quantumalaviya replied:
`MixUp` is implemented in a similar way.

bhack replied:
Oh yes it is seems that it was introduced previously also in `MIxUp` by @qlzh727 with https://github.com/keras-team/keras-cv/commit/2c42c48bc6a449799fa008f1a8a808ee1952c75c

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,221 @@
+        return tf.math.sqrt(fx * fx + fy * fy)
+
+    def get_spectrum(self, freqs, decay_power, channel, h, w):
+        # Function to apply a low pass filter by decaying its high frequency components.
```
Let's try to remove blanklines to create logical blocks

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,221 @@
+
+    def binarise_mask(self, mask, lam, in_shape, edge_softening=0.0):
+        # Create the final mask from the sampled values.
+
```
Let's remove these newlines after comments.  No need for them, other than that the newlines are great!

LukeWood commented:
I approved these changes, but I'm still not convinced of the name `FMix`.  I don't think the naming conventions of research are a strong enough signal to drive our naming as they often change over the years.  There are some cases where I believe the technique names are accurate enough to keep (i.e. RandAugment), but `FMix` is very generic and I don't think particularly clear.  FourierMix is clearer to me.

Thoughts on naming here @fchollet @qlzh727?

This layer performs CutMix using shapes sampled from fourier (aka frequency) domain.  To me, FourierMix sounds good, but I worry `Fourier.*` may become too generic as time goes on.  Perhaps `FMix` is actually good for this reason.

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,221 @@
+        return sample_alpha / (sample_alpha + sample_beta)
+
+    @staticmethod
+    def fftfreq(signal_size, sample_spacing=1):
```
lets make this private.

_fftfreq

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,221 @@
+
+        return results / (signal_size * sample_spacing)
+
+    def apply_fftfreq(self, h, w):
```
make private

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,221 @@
+
+        return tf.math.sqrt(fx * fx + fy * fy)
+
+    def get_spectrum(self, freqs, decay_power, channel, h, w):
```
same here - make private

LukeWood commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,221 @@
+
+        return scale * param
+
+    def sample_mask_from_transform(self, decay, shape, ch=1):
```
etc for the remaining helper methods.  We don't want these showing up on the public API Docs.

LukeWood commented:
> @MattPainter01

Would be great to get your code review if possible!

innat commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,221 @@
+
+        fx = FMix.fftfreq(w)[: w // 2 + 1 + w % 2]
+
+        fy = FMix.fftfreq(h)
```
@quantumalaviya linebreak in 79 isn't necessary, is it?

quantumalaviya replied:
Originally added it to separate the calculation of `fx` and `fy`. Do you suggest I remove it?

innat replied:
I think so, not necessary to line break here.

innat commented on keras_cv/layers/preprocessing/fmix.py:
```diff
@@ -0,0 +1,221 @@
+
+    def _fmix(self, images):
+        shape = tf.shape(images)
+        permutation_order = tf.random.shuffle(tf.range(0, shape[0]), seed=self.seed)
```
@quantumalaviya In order to reproduce the o/p with a given seed value, I think it's needed to do as follows:

```python
tf.random.set_seed(seed=self.seed)
tf.random.shuffle(tf.range(0, shape[0]), seed=self.seed)
```
Can u please check?

quantumalaviya replied:
The `BaseImageAugmentationLayer` takes care of this, I believe. It inherits the `BaseRandomLayer` to which the seed is passed.

sebastian-sz replied:
@quantumalaviya I think that to use that `seed` you would need to replace `tf.random` ops with `self._random_generator`.

Looking at the code of [BaseRandomLayer](https://github.com/keras-team/keras/blob/275a0707c066bd90501e3fb3a0c26b1b676c06d3/keras/engine/base_layer.py#L3342)
```python
self._random_generator = backend.RandomGenerator(
        seed, force_generator=force_generator)
```

quantumalaviya replied:
Yes, that actually makes sense @sebastian-sz. I didn't realize I haven't used the `_random_generator` everywhere.

Correct me if I am wrong here but `backend.RandomGenerator` doesn't provide a shuffle function (although it claims it should be used for shuffling). Am I missing something here?

sebastian-sz replied:
@quantumalaviya Yes you are right - this particular op does not seem to have Generator equivalent.

quantumalaviya replied:
I believe @innat's approach should work then:

```python
tf.random.set_seed(seed=self.seed)
tf.random.shuffle(tf.range(0, shape[0]), seed=self.seed)
```

LukeWood replied:
I don't think we want to be setting a seed inside of call like this actually.

LukeWood replied:
For now, feel free to just pass `seed` to the `tf.random.shuffle` as you have done.

MattPainter01 commented:
This is really cool work! 

The implementation looks good to me, but unfortunately I don't have much experience with tf, so I can't really comment on the specifics... Sorry! 

Just a couple of high level comments:
- From what I remember, we found that the mask softening resulted in no real benefits for image tasks, and sometimes hindered learning. You might want to consider removing it, or adding a disclaimer to the comments. We did find that it helped somewhat on text tasks, but presumably that's not within the keras-cv domain. 
- Whilst you didn't implement it here, just thought I'd mention that the reformulation we tried also didn't help. So I wouldn't put any effort into implementing/allowing that either. The reformulation I refer to is [here](https://github.com/ecs-vlc/FMix/blob/e5991dca018882734c8ea63599f10dfbe67fa0ae/fmix.py#L81) in our repo, and I think was discussed in the v1 arxiv paper.
- I notice you're sticking to 2D? I see the early focus of keras-cv is images, but I wonder if it's planned to include video eventually?  I don't think we explored 3D FMix for video too much, but maybe it's something to consider for the future.


As for the name, the F in FMix was meant to stand for Fourier, as you all guessed. It was an oversight that we didn't include that somewhere in the paper. That being said, do feel free to rename your implementation to whatever best fits your repo/structure!

quantumalaviya commented:
Thanks for the review @MattPainter01! 

@LukeWood Which of these changes do you suggest we adopt? 
As for the 3D support, I got rid of it after the commit to support varying masks. Although it may be desirable to support videos, are we focusing on that right now? If so, I don't mind making some changes to reintroduce it!

LukeWood commented:
> This is really cool work!
> 
> The implementation looks good to me, but unfortunately I don't have much experience with tf, so I can't really comment on the specifics... Sorry!
> 
> Just a couple of high level comments:
> 
> * From what I remember, we found that the mask softening resulted in no real benefits for image tasks, and sometimes hindered learning. You might want to consider removing it, or adding a disclaimer to the comments. We did find that it helped somewhat on text tasks, but presumably that's not within the keras-cv domain.
> * Whilst you didn't implement it here, just thought I'd mention that the reformulation we tried also didn't help. So I wouldn't put any effort into implementing/allowing that either. The reformulation I refer to is [here](https://github.com/ecs-vlc/FMix/blob/e5991dca018882734c8ea63599f10dfbe67fa0ae/fmix.py#L81) in our repo, and I think was discussed in the v1 arxiv paper.
> * I notice you're sticking to 2D? I see the early focus of keras-cv is images, but I wonder if it's planned to include video eventually?  I don't think we explored 3D FMix for video too much, but maybe it's something to consider for the future.
> 
> As for the name, the F in FMix was meant to stand for Fourier, as you all guessed. It was an oversight that we didn't include that somewhere in the paper. That being said, do feel free to rename your implementation to whatever best fits your repo/structure!

Thanks for your detailed review @MattPainter01.  I really appreciate it. 

I think I am more comfortable with the name `FourierMix` or `FourierCutMix`.  I'm trying to think if `FourierMix` could ever have a naming collision... but I think it is safe to use here.

`FMix` just doesn't match the other layers or the Keras ecosystem.  `FourierMix` also gives some context to users on the behavior.  Thanks

@quantumalaviya FYI

LukeWood commented:
> Thanks for the review @MattPainter01!
> 
> @LukeWood Which of these changes do you suggest we adopt? As for the 3D support, I got rid of it after the commit to support varying masks. Although it may be desirable to support videos, are we focusing on that right now? If so, I don't mind making some changes to reintroduce it!

3D will likely have to come at the BaseImageAugmentationLayer down the line.  No point in adding it to one layer for now; lets do it as a big effort down the line where we can support it from ALL layers

LukeWood commented:
> From what I remember, we found that the mask softening resulted in no real benefits for image tasks, and sometimes hindered learning. You might want to consider removing it, or adding a disclaimer to the comments. We did find that it helped somewhat on text tasks, but presumably that's not within the keras-cv domain.

Actually, @quantumalaviya I missed this point on my comment pass.  Should we remove softening?  I'd be in favor if the author @MattPainter01 found no benefit.

innat commented on keras_cv/layers/preprocessing/fourier_mix.py:
```diff
@@ -182,9 +165,7 @@ def _fourier_mix(self, images):
 
         # binarise masks utilizing mapped calls
         masks = tf.map_fn(
```
@quantumalaviya is it possible to use `tf.vectorized_map` instead?

quantumalaviya replied:
I was just wondering this. I am running the experiment rn!

LukeWood replied:
Much better.  Performance comes second to correctness.

quantumalaviya commented:
@LukeWood mask softening has also been removed btw!

LukeWood commented:
Thank you for the wonderful contribution @quantumalaviya!  Looks good, I'll merge this following minor copyedits!
